### ISSUE TITLE
STM32F103 support

### ISSUE THREAD

**Issue by etet100** [Rejdok/STM32-Marlin-Port#2]
Hi. Could you tell me how hard would it be to use STM32F103 (maple mini) instead of F4xx ?

**Comment by Rejdok** [comment]
Hello. I think it wouldn't very difficult. You should change stm hal from F4xx series to F1xx and change timers frequency.

**Comment by etet100** [comment]
Hi. Thank you for your answer. It is quite hard actually. After a few hours o trials i have 463 errors left.

**Comment by Rejdok** [comment]
Hello. 463 errors uts too much. I hope you build it using sw4stm32. Just create new build, from debug, then change preprocessor definition in project properties stm32f407xx to stm32f103xB or other. And exclude from build hal drivers for stm32f407. Then change linker script in project settings, to ls for your's microcontroller. You can simply generate it using stm32cubeMX. And one more step exclude from build  
startup_stm32f407xx.s, then add startup file for yours mc.

**Comment by etet100** [comment]
It's hopeless, i'm affraid. Hundreds of errors. I give up.

**Comment by Rejdok** [comment]
Ok, i add stm32f103c8t6 support. At least it builded successful . Can you change timers frequency in marlin/portable (stm32_stepper.cpp and stm32_temperature.cpp and remap pins for your needs? So, if  you get some errors during debug, you can write here.

**Comment by etet100** [comment]
Thank you very much. Unfortunatelly, i'm still experiencing a lot of issues. Some include paths are missing (fatal error: stm32f1xx.h: No such file or directory), floating point ABI is set to hard, ARM_MATH_CM4 is defined instead of ARM_MATH_CM3 (for Cortex-M3 target), ADC driver seems to be missing (no stm32f1xx_hal_adc.h), USB driver too (provided one uses USB OTG mode which is not supported by STM32F103CB). I had to change some includes (#include "stm32f4xx.h" to #include "stm32f1xx.h"). And now i have more and more errors. I have  no idea what i'm doing wrong. Are you sure you are able to compile it without errors?

**Comment by Rejdok** [comment]
Yep. Sorry, my bad. I'm overlooked little bit. Try it now=)  Chose pins, configure it using hal, then remap pins in marlin/portable/port.cpp, and change timers frequency.

**Comment by etet100** [comment]
Thanks. Now it freezes on:

```
while(hUsbDeviceFS.dev_state!=3)
  HAL_Delay(500);
```
dev_state is always = 1

Interestingly, microcontroller also hangs when i upload
freshly created Stm32CubeMX project. Damn. I'm sure my 
maple mini board is perfectly fine.

**Comment by Rejdok** [comment]
Yep. It's ok. Just connect via USB your mapple mini and PC.

**Comment by Rejdok** [comment]
And mc dosht hangs. Just enable debug by serial wire in sys tab.

**Comment by etet100** [comment]
It is connected to PC. And serial wire debugging is enabled. I see that USB
is not working properly:

![maple_hangs_1](https://user-images.githubusercontent.com/22624537/36947251-a90e52c8-1fc9-11e8-9dfc-740ab34dd9b4.png)

It falls into infinite loop, every time.

**Comment by Rejdok** [comment]
That's strange. It's screenshot from custom project? Did you configure usb device as virtual com port in cubeMx, and  usb clock configuration? Your pc detect new device?

**Comment by etet100** [comment]
This is minimal example to show how it works. Marlin code behave exactly the same. Clock settings are copied directly from code generated by CubeMx. No new device is detected by pc. 

![maple_hangs_2](https://user-images.githubusercontent.com/22624537/36949013-2243d852-1fe3-11e8-8fcf-6b9da268808b.png)

**Comment by Rejdok** [comment]
Hm, I'm remembered. State changes, when you connect to com port using terminal. I found mc stm32f103rc, and seen, that communication doesn't work. So I'm fix it now.  Sorry for troubles. I wrote this port about year ago.
But at my mc USB COM port works. And pc found it. Try check COM port section in device manager.

**Comment by Rejdok** [comment]
Ok now it works as expected.

**Comment by Rejdok** [comment]
And if device not detected, try to change USB cable. =) Maybe it would work=)

**Comment by etet100** [comment]
And still no connection. I have no idea what is going on. I have checked everything many times. Maple board is fine (it's working perfectly when i upload original maple firmware), cables are fine, firmware seems to be fine.

**Comment by Rejdok** [comment]
https://os.mbed.com/media/uploads/hudakz/maple_usb.png
https://os.mbed.com/users/hudakz/code/MapleMini_USBSerial/rev/d37938a78bc5/
https://en.wikipedia.org/wiki/USB see line state
![](https://os.mbed.com/media/uploads/hudakz/maple_usb.png)

Try to set PB9 in high level.

**Comment by etet100** [comment]
You are right! I have replaced Q2 with bridge. Of course it doesn't mean that everything is working. Communication is not stable, motors are not working, timers are not working. I found a few bugs.

**Comment by Rejdok** [comment]
Timers work's now. 
Communications should be stabile.

**Comment by bhgv** [comment]
hi, i send you a PR with some editings and addings. mostly for maple mini. README.md and a pair of alternative ROM uploaders for people without stlink.

you may see how it looks here:
https://github.com/bhgv/STM32-Marlin-Port

**Comment by etet100** [comment]
Hi. I wonder what you think about using more recent Marlin v2 which introduces hardware abstraction layer.

**Comment by bhgv** [comment]
@etet100 
Give me please a link to v2

Is it your https://github.com/etet100/marlin_cetus_3d ?

**Comment by Rejdok** [comment]
https://github.com/MarlinFirmware/Marlin/tree/bugfix-2.0.x . In Marlin v2 developers made a lot of great changes in code style and project organisation. But at present this branch unstable and actively develop. So they can easily change functions definitions, project structure e.t.c. That why I dont want using Marlin v2 now. In fact, if I had more time, I will port smootheware on stm32, and use this firmware. Sorry for bad english.
